### PR TITLE
fix: Disallow non null updates for transactions

### PR DESCRIPTION
**Pull request by Artemka374**
## What ❔

Add check to `set_tx_id` query, that doesn't allow to update tx_id if it is not NULL.

## Why ❔

To prevent issues when 1 batch can be commited 2 times.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [x] Spellcheck has been run via `zk spellcheck`.

### CODE REVIEW

File: core/lib/dal/src/blocks_dal.rs [hunk 1]
```diff
@@ -402,17 +402,18 @@ impl BlocksDal<'_, '_> {
         number_range: ops::RangeInclusive<L1BatchNumber>,
         eth_tx_id: u32,
         aggregation_type: AggregatedActionType,
-    ) -> DalResult<()> {
+    ) -> anyhow::Result<()> {
```
Comment (Deniallugo) [inline core/lib/dal/src/blocks_dal.rs:405]: please keep dal error

**Review by Deniallugo — CHANGES_REQUESTED** [review]
Please add tests for this feature

File: core/lib/dal/src/blocks_dal.rs (hunk 1 above)
Comment (Artemka374) [reply core/lib/dal/src/blocks_dal.rs:405]: Well... I can't. There is no possibility to customize the DalResult. So it's either anyhow or panic

Comment (Deniallugo) [reply core/lib/dal/src/blocks_dal.rs:405]: @slowli what is the best approach for it? 
Probably we could add custom errors to dal error?

**Review by RomanBrodetski — DISMISSED** [review]
very nice!

File: core/lib/dal/src/blocks_dal.rs (hunk 1 above)
Comment (montekki) [reply core/lib/dal/src/blocks_dal.rs:405]: maybe it makes sense panicking here as we've discussed it should panic anyway.

Comment (Deniallugo) [reply core/lib/dal/src/blocks_dal.rs:405]: Just for me, it's better to panic, than using anyhow... 
But i think we have an ability to customize the errors

Comment (perekopskiy) [reply core/lib/dal/src/blocks_dal.rs:405]: `Instrumented::constraint_error` fits here

Comment (Deniallugo) [reply core/lib/dal/src/blocks_dal.rs:405]: Oh, thank you @perekopskiy! That's what I wanted

File: core/lib/dal/src/blocks_dal.rs [hunk 2]
```diff
@@ -423,16 +424,30 @@ impl BlocksDal<'_, '_> {
                 .with_arg("eth_tx_id", &eth_tx_id)
                 .execute(self.storage)
                 .await?;
+
+                if result.rows_affected() == 0 {
+                    let err = Instrumented::new("set_eth_tx_id#commit")
+                        .with_arg("number_range", &number_range)
+                        .with_arg("eth_tx_id", &eth_tx_id)
+                        .arg_error(
```
Comment (slowli) [inline core/lib/dal/src/blocks_dal.rs:432]: Nit: This is a constraint error rather an arg one (an arg error should be possible to produce without running DB queries).

File: core/lib/dal/src/blocks_dal.rs [hunk 3]
```diff
@@ -423,16 +424,30 @@ impl BlocksDal<'_, '_> {
                 .with_arg("eth_tx_id", &eth_tx_id)
                 .execute(self.storage)
                 .await?;
+
+                if result.rows_affected() == 0 {
+                    let err = Instrumented::new("set_eth_tx_id#commit")
```
Comment (slowli) [inline core/lib/dal/src/blocks_dal.rs:429]: Nit: You can define this `Instrumented` instance before the query and share it both for the query (via `Instrumented::with()`) and the following check. Would require minor changes to the `db_connection` crate (namely, adding `Clone` for `Instumented<'_, ()>`).

**Review by Deniallugo — APPROVED** [review]

**Review by slowli — APPROVED** [review]

**Review by montekki — APPROVED** [review]

**Pull request merged**